### PR TITLE
feat:add server function in RpcServerArgs.

### LINKED ISSUE #1287: Configure RpcServer from RpcServerArgs
### Describe the feature

`RpcServerArgs`contains all settings for the RPC server

https://github.com/paradigmxyz/reth/blob/afb6671a0f8c681e1fc3b3d13d6a33f30c954dd2/bin/reth/src/args/rpc_server_args.rs#L12

the RPC server needs to be configured from the RpcServerArgs

## TODO
* essentially move this section https://github.com/paradigmxyz/reth/blob/afb6671a0f8c681e1fc3b3d13d6a33f30c954dd2/bin/reth/src/node/mod.rs#L143-L151 to RpcServerArgs
* add `RpcServerArgs::server` function that returns `RpcServer`, same trait bounds as the `launch` function

server builder needs some changes so it mirrors the cli args: ref #1231

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by GrapeBaBa**
Fix #1287

### CODE REVIEW

File: bin/reth/src/args/rpc_server_args.rs [hunk 1]
```diff
@@ -78,6 +84,27 @@ impl RpcServerArgs {
             }
         }
     }
+
+    pub(crate) async fn server<Client, Pool, Network>(
```
Comment (gakonst) [inline bin/reth/src/args/rpc_server_args.rs:88]:
```suggestion
    pub(crate) async fn server<Client, Pool, Network>(
        &self,
```

File: bin/reth/src/args/rpc_server_args.rs [hunk 2]
```diff
@@ -78,6 +84,27 @@ impl RpcServerArgs {
...
+        Pool: TransactionPool + Clone + 'static,
+        Network: NetworkInfo + Peers + Clone + 'static,
+    {
+        reth_rpc_builder::launch(
+            client,
+            pool,
+            network,
+            TransportRpcModuleConfig::default()
+                .with_http(vec![RethRpcModule::Admin, RethRpcModule::Eth]),
+            RpcServerConfig::default().with_http(Default::default()),
```
Comment (gakonst) [inline bin/reth/src/args/rpc_server_args.rs:104]: use the options from `self.http` / `self.ws` etc. to configure these

File: bin/reth/src/node/mod.rs [hunk 3]
```diff
@@ -149,13 +149,10 @@ impl Command {
         let _secret = self.rpc.jwt_secret();
 
         // TODO(mattsse): cleanup, add cli args
-        let _rpc_server = reth_rpc_builder::launch(
+        let _rpc_server = RpcServerArgs::server(
```
Comment (gakonst) [inline bin/reth/src/node/mod.rs:152]:
```suggestion
        let _rpc_server = self.rpcserver(
```

File: bin/reth/src/node/mod.rs [hunk 4]
```diff
@@ -149,13 +149,10 @@ impl Command {
         let _secret = self.rpc.jwt_secret();
```
Comment (gakonst) [inline bin/reth/src/node/mod.rs:149]: @rkrasiuk i think in one of your PRs we need to actually spawn the engine api here

**Review by gakonst — CHANGES_REQUESTED** [review]

File: bin/reth/src/args/rpc_server_args.rs [hunk 5]
```diff
@@ -99,12 +105,60 @@ impl RpcServerArgs {
...
+            config = config.with_ipc(ipc_builder).with_ipc_endpoint(
+                self.ipcpath.as_ref().unwrap_or(&DEFAULT_IPC_ENDPOINT.to_string()),
+            );
+        }
+
+        RpcServerConfig::from(config)
+    }
+
+    #[allow(dead_code)]
+    async fn rpc_server(&self) -> Result<RpcServer, RpcError> {
```
Comment (GrapeBaBa) [inline bin/reth/src/args/rpc_server_args.rs:159]: @mattsse I am not sure why need a function return `RpcServer`, it is just delegate `build` function.

File: bin/reth/src/args/rpc_server_args.rs [hunk 6]
```diff
@@ -99,12 +105,60 @@ impl RpcServerArgs {
             client,
```
Comment (GrapeBaBa) [inline bin/reth/src/args/rpc_server_args.rs:105]: @gakonst @mattsse for `server` function what is your thought? If just call `launch` function or actually need reimplement similar logic in `RpcServerArgs`?

File: crates/rpc/rpc-builder/src/lib.rs [hunk 7]
```diff
@@ -636,7 +636,7 @@ impl RpcServerConfig {
 #[derive(Debug, Clone, Default, Eq, PartialEq)]
 pub struct TransportRpcModuleConfig {
     /// http module configuration
-    http: Option<RpcModuleSelection>,
+    pub http: Option<RpcModuleSelection>,
```
Comment (GrapeBaBa) [inline crates/rpc/rpc-builder/src/lib.rs:639]: @gakonst @mattsse I want to create some tests, just make field pub or add query method?

Comment (gakonst) [reply crates/rpc/rpc-builder/src/lib.rs:639]: add method

File: bin/reth/src/node/mod.rs [hunk 8]
```diff
@@ -149,15 +149,14 @@ impl Command {
...
-        .await?;
+        let _rpc_server = self
+            .rpc
+            .server(
+                ShareableDatabase::new(db.clone(), self.chain.clone()),
+                reth_transaction_pool::test_utils::testing_pool(),
+                network.clone(),
+            )
+            .await?;
         info!(target: "reth::cli", "Started RPC server");
```
Comment (gakonst) [inline bin/reth/src/node/mod.rs:160]: much nicer. can remove the todo

File: bin/reth/src/args/rpc_server_args.rs (hunk 5 above)
Comment (gakonst) [reply bin/reth/src/args/rpc_server_args.rs:159]: i think remove given the server fn above?

File: bin/reth/src/args/rpc_server_args.rs [hunk 9]
```diff
@@ -78,6 +89,76 @@ impl RpcServerArgs {
...
+            self.transport_rpc_module_config(),
+            self.rpc_server_config(),
+        )
+        .await
+    }
+
+    fn transport_rpc_module_config(&self) -> TransportRpcModuleConfig {
+        let mut config = TransportRpcModuleConfig::default();
+        let rpc_modules =
+            RpcModuleSelection::Selection(vec![RethRpcModule::Admin, RethRpcModule::Eth]);
```
Comment (gakonst) [inline bin/reth/src/args/rpc_server_args.rs:117]: for a follow-up, but we should also add these to the cli config

Comment (GrapeBaBa) [reply bin/reth/src/args/rpc_server_args.rs:117]: i can't get you.

File: crates/rpc/rpc-builder/src/lib.rs (hunk 7 above)
Comment (mattsse) [reply crates/rpc/rpc-builder/src/lib.rs:639]: yes, please add functions instead.

File: bin/reth/src/args/rpc_server_args.rs (hunk 5 above)
Comment (mattsse) [reply bin/reth/src/args/rpc_server_args.rs:159]: this is still useful as a convenient function.

File: bin/reth/src/args/rpc_server_args.rs [hunk 10]
```diff
@@ -78,6 +89,76 @@ impl RpcServerArgs {
...
+            config = config.with_ipc(ipc_builder).with_ipc_endpoint(
+                self.ipcpath.as_ref().unwrap_or(&DEFAULT_IPC_ENDPOINT.to_string()),
+            );
+        }
+
+        config
+    }
+
+    #[allow(dead_code)]
+    async fn rpc_server(&self) -> Result<RpcServer, RpcError> {
```
Comment (mattsse) [inline bin/reth/src/args/rpc_server_args.rs:159]: please add docs
```suggestion
    
    pub async fn rpc_server(&self) -> Result<RpcServer, RpcError> {
```

File: bin/reth/src/args/rpc_server_args.rs [hunk 11]
```diff
@@ -78,6 +89,76 @@ impl RpcServerArgs {
             }
         }
     }
+
+    pub(crate) async fn server<Client, Pool, Network>(
```
Comment (mattsse) [inline bin/reth/src/args/rpc_server_args.rs:93]: this needs docs which include the info that it spawns the server and returns the handle. See launch docs.

and should be named `start_server` or `launch_server` instead.

File: bin/reth/src/node/mod.rs [hunk 12]
```diff
@@ -149,15 +149,14 @@ impl Command {
         let _secret = self.rpc.jwt_secret();
 
         // TODO(mattsse): cleanup, add cli args
```
Comment (mattsse) [inline bin/reth/src/node/mod.rs:151]: can be removed now

**Review by mattsse — CHANGES_REQUESTED** [review]
looks great.

functions need some docs, otherwise lgtm.
some suggestions re naming.

ty for the tests!

File: bin/reth/src/args/rpc_server_args.rs (hunk 5 above)
Comment (GrapeBaBa) [reply bin/reth/src/args/rpc_server_args.rs]: how about add back when using it?

File: crates/rpc/rpc-builder/src/lib.rs [hunk 13]
```diff
@@ -683,6 +698,21 @@ impl TransportRpcModuleConfig {
...
+    }
+
+    /// Returns the [RpcModuleSelection] for the ws transport
+    pub fn ws(&self) -> Option<RpcModuleSelection> {
+        self.ws.clone()
+    }
+
+    /// Returns the [RpcModuleSelection] for the ipc transport
+    pub fn ipc(&self) -> Option<RpcModuleSelection> {
+        self.ipc.clone()
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:714]: last nit:

we should return references here.

**Review by mattsse — CHANGES_REQUESTED** [review]
one last nit.

File: bin/reth/src/args/rpc_server_args.rs (hunk 9 above)
Comment (gakonst) [reply bin/reth/src/args/rpc_server_args.rs:118]: I mean that Admin & Eth should not always be enabled by default, the user should specify which apis they want to enable. 

@mattsse I also wonder if we should ship some solution for having private / public APIs, e.g. exposing the Admin API on a separate port that is not exposed to the internet

**Review by gakonst — APPROVED** [review]

**Pull request merged**